python/ql/src/experimental/dataflow/internal/Attributes.qll

```diff
@@ -0,0 +1,208 @@
+private class SetAttrCallNode extends BuiltinAttrCallNode {
+  SetAttrCallNode() { name = "setattr" }
+}
+
+/** Represents calls to the built-in `getattr`. */
+private class GetAttrCallNode extends BuiltinAttrCallNode {
+  GetAttrCallNode() { name = "getattr" }
+}
+
+/** An attribute assignment using `setattr`, e.g. `setattr(object, attr, value)` */
+private class SetAttrCallAsAttrWrite extends AttrWrite, CfgNode {
+  SetAttrCallNode setattr_call;
+
+  SetAttrCallAsAttrWrite() { this = TCfgNode(setattr_call) }
+
+  override Node getValue() { result = TCfgNode(setattr_call.getValue()) }
+
+  override Node getObject() { result = TCfgNode(setattr_call.getObject()) }
+
+  override ExprNode getAttributeNameExpr() { result = TCfgNode(setattr_call.getName()) }
+
+  override string getAttributeName() {
+    // TODO track this back using local flow
+    exists(StrConst s, Node nodeFrom |
+      s = nodeFrom.asExpr() and
+      simpleLocalFlowStep*(nodeFrom, this.getAttributeNameExpr()) and
+      result = s.getText()
+    )
+  }
+}
+
+/**
+ * An attribute assignment via a class field, e.g.
+ * ```python
+ * class MyClass:
+ *     attr = value
+ * ```
+ * is treated as equivalent to `MyClass.attr = value`.
+ */
+private class ClassDefinitionAsAttrWrite extends AttrWrite, Node {
+  ClassExpr cls;
+  DefinitionNode attr_node;
+
+  ClassDefinitionAsAttrWrite() {
+    attr_node instanceof NameNode and
+    this.asCfgNode() = attr_node and
+    attr_node.getScope() = cls.getInnerScope()
+  }
+
+  override Node getValue() { result = TCfgNode(attr_node.getValue()) }
+
+  override Node getObject() { result = TCfgNode(cls.getAFlowNode()) }
+
+  override ExprNode getAttributeNameExpr() { none() }
+
+  override string getAttributeName() { result = attr_node.(NameNode).getId() }
+}
+
+/**
+ * A read of an attribute on an object. This includes
+ * - Simple attribute reads: `object.attr`
+ * - Dynamic attribute reads using `getattr`: `getattr(object, attr)`
+ * - Qualified imports: `from module import attr as name`
+ */
+abstract class AttrRead extends AttrRef, Node { }
+
+/** A simple attribute read, e.g. `object.attr` */
+private class AttributeReadAsAttrRead extends AttrRead, CfgNode {
+  AttrNode attr_node;
+
+  AttributeReadAsAttrRead() { this = TCfgNode(attr_node) }
+
+  override Node getObject() { result = TCfgNode(attr_node.getObject()) }
+
+  override ExprNode getAttributeNameExpr() {
+    // Attribute names don't exist as `Node`s in the control flow graph, as they can only ever be
+    // identifiers, and are therefore represented directly as strings.
+    // Use `getAttributeName` to access the name of the attribute.
+    none()
+  }
+
+  override string getAttributeName() { result = attr_node.getName() }
+}
+
+/** An attribute read using `getattr`: `getattr(object, attr)` */
+private class GetAttrCallAsAttrRead extends AttrRead, CfgNode {
+  GetAttrCallNode getattr_call;
+
+  GetAttrCallAsAttrRead() { this.asCfgNode() = getattr_call }
```

Comment:
I think this can be written
`GetAttrCallAsAttrRead() { node instanceof GetAttrCallNode }`
and then you do not need the field `getattr_call` (but can simply use the field `node` inherited from `CfgNode`).

Reply:
Hmm... I thought I tried a variation on this, but couldn't get it to do the right thing. I _did_ notice a similar pattern in the JS libraries (involving overriding of fields), so I'll see if I can figure this out.

Reply:
Ah, now I remember the issue. If I do
```ql
override GetAttrCallNode node;
```
I get an error stating that `GetAttrCallNode` is not a subtype of `DataFlowCfgNode`. I can instead do what you suggest, without the `override` in place, but then I would have to explicitly cast to `GetAttrCallNode` before accessing its member methods (since `node` would just be a generic `DataFlowCfgNode`). Perhaps I should be extending `DataFlowCfgNode` instead of `ControlFlowNode` for my built-ins? That seems to be the only way out of this mess. I think I'll try this.

Reply:
This turned out to be fairly straight forward to do, with the right inheritance structure.